helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java

```diff
@@ -92,19 +98,42 @@ public ZkBaseDataAccessor(HelixZkClient zkClient) {
     _zkClient = zkClient;
   }
 
+  private synchronized HelixZkClient getNonZNRecordClient() {
+    if (_nonZNRecordClient == null) {
+      _nonZNRecordClient = new ZkClient.Builder().setConnection(
+          new ZkConnection(_zkClient.getServers(), ZkClient.DEFAULT_SESSION_TIMEOUT))
+          .setZkSerializer(new ZkSerializer() {
+            @Override
+            public byte[] serialize(Object data) throws ZkMarshallingError {
+              if (data instanceof byte[]) {
+                return (byte[]) data;
+              }
+              throw new HelixException("Only support a byte array as an argument!");
+            }
+
+            @Override
+            public Object deserialize(byte[] data) throws ZkMarshallingError {
+              return data;
+            }
+          }).build();
+    }
+
+    return _nonZNRecordClient;
+  }
+
   /**
    * sync create
    */
   @Override
   public boolean create(String path, T record, int options) {
-    AccessResult result = doCreate(path, record, options);
+    AccessResult result = doCreate(path, record, options, true);
     return result._retCode == RetCode.OK;
   }
 
   /**
-   * sync create
+   * sync create. In order to support non-ZnRecord creation, the instance of {@link HelixZkClient} needs to be a parameter
    */
-  public AccessResult doCreate(String path, T record, int options) {
+  public AccessResult doCreate(String path, Object record, int options, boolean isZnRecord) {
```

Comment:
Could we possibly tweak the method signature like it was before instead of a boolean flag? I think

`doCreate(HelixZkClient zkClient, Object record, int options);` was a better structure.

And getNonZnRecordZkClient() would be called in your get(), set(), and create().

Reply:
However, it will go back to the previous problem when ZKCachedDataAccesor wants to use `doCreate` method directly and it needs to get the instance of ZKClient.

Reply:
CachedBaseDataAccessor could just pass in `_zkclient`, correct?

Reply:
ZkCachedBaseDataAccessor doesn't have the `_zkClient` instance, ZkCachedBaseDataAccessor is not a child class of zkBaseDataAccessor somehow. Going this route will create the public method of getting zkClient.

Reply:
I take the comment back, sorry I just noticed the cacheAccessor also has its own instance of zkClient.